src/compartment.c

```diff
@@ -707,79 +882,55 @@ comp_clean(struct Compartment *to_clean)
 static ssize_t
 do_pread(int fd, void *buf, size_t count, off_t offset)
 {
-    size_t res = pread(fd, buf, count, offset);
+    ssize_t res = pread(fd, buf, count, offset);
     if (res == -1)
     {
         err(1, "Error in pread");
     }
     return res;
 }
 
-static Elf64_Sym *
-find_symbols(const char **names, size_t names_to_find_count, bool find_all,
-    Elf64_Sym *symtb, char *strtb, size_t symtb_sz)
+static void
+get_lib_name(struct LibDependency *lib_dep, const char *lib_path)
 {
-    Elf64_Sym *found_syms = calloc(names_to_find_count, sizeof(Elf64_Sym));
-    Elf64_Sym curr_sym;
-    size_t found_syms_count = 0;
-    for (size_t i = 0; i < symtb_sz / sizeof(Elf64_Sym); ++i)
-    {
-        curr_sym = symtb[i];
-        for (size_t j = 0; j < names_to_find_count; ++j)
-        {
-            // XXX As a follow-up from how we handle the string table, here we
-            // get symbol names by indexing at the `char` offset, then getting
-            // the string pointer (equivalent to `strtb + curr_sym.st_name`).
-            if (!strcmp(names[j], &strtb[curr_sym.st_name]))
-            {
-                found_syms[j] = curr_sym;
-                found_syms_count += 1;
-            }
-        }
-    }
+    const char *basename = strrchr(lib_path, '/') + 1;
+    lib_dep->lib_name = malloc(strlen(basename));
+    strcpy(lib_dep->lib_name, basename);
+}
+
+static void *
+extract_sym_offset(struct Compartment *comp, struct LibSymSearchResult res)
+{
+    return (char *) comp->libs[res.lib_idx]->lib_mem_base
+        + (intptr_t) comp->libs[res.lib_idx]->lib_syms[res.sym_idx].sym_offset;
+}
 
-    // If we didn't find all symbols that we wanted to intercept, throw an error
-    if (find_all && found_syms_count != names_to_find_count)
+static struct LibSymSearchResult
+find_lib_dep_sym_in_comp(
+    const char *to_find, struct Compartment *comp_to_search)
+{
+    for (size_t i = 0; i < comp_to_search->libs_count; ++i)
     {
-        const char **not_found_syms = malloc(names_to_find_count);
-        size_t not_found_idx = 0;
-        for (size_t i = 0; i < names_to_find_count; ++i)
+        for (size_t j = 0; j < comp_to_search->libs[i]->lib_syms_count; ++j)
         {
-            bool not_found = true;
-            for (size_t j = 0; j < found_syms_count; ++j)
-            {
-                if (!strcmp(&strtb[found_syms[j].st_name], names[i]))
-                {
-                    not_found = false;
-                    break;
-                }
-            }
-            if (not_found)
+            if (!strcmp(to_find, comp_to_search->libs[i]->lib_syms[j].sym_name))
             {
-                not_found_syms[not_found_idx] = names[i];
-                not_found_idx += 1;
+                struct LibSymSearchResult res = { i, j };
+                return res;
```

Comment:
I *think* you might be able to do `return struct LibSymSearchResult { i, j }`.

Reply:
Doesn't seem so:
```
In file included from ../tests/test_map.c:1:
../src/compartment.c:917:24: error: expected expression
                return struct LibSymSearchResult { i, j };
                       ^
../src/compartment.c:921:12: error: expected expression
    return struct LibSymSearchResult { -1, -1 };
           ^
2 errors generated.
```